src/libraries/System.Private.Uri/src/System/IriHelper.cs

```diff
@@ -67,114 +93,55 @@ internal static bool CheckIsReserved(char ch, UriComponents component)
         internal static unsafe string EscapeUnescapeIri(char* pInput, int start, int end, UriComponents component)
         {
             int size = end - start;
-            ValueStringBuilder dest = new ValueStringBuilder(size);
-            byte[]? bytes = null;
-
-            int next = start;
-            char ch;
+            ValueStringBuilder dest = size <= 256
+                ? new ValueStringBuilder(stackalloc char[256])
+                : new ValueStringBuilder(size);
 
             Span<byte> maxUtf8EncodedSpan = stackalloc byte[4];
 
-            for (; next < end; ++next)
+            for (int i = start; i < end; ++i)
             {
-                if ((ch = pInput[next]) == '%')
+                char ch = pInput[i];
+                if (ch == '%')
                 {
-                    if (next + 2 < end)
+                    if (i + 2 < end)
```

Comment:
Technically this addition could be subject to integer overflow with large-enough input indexes. (This isn't a bug introduced in this PR; it's an existing issue.) This isn't a problem at this exact moment since the caller limits string length, but it's something that we should consider addressing now so that it doesn't become an issue in a future world where the caller allows larger values for _start_ and _end_.

The correct, non-overflowing way of performing this check is `if (end - i < 2)`.

Reply:
Typo, it should be `if (end - i > 2)`.